[WebGPU] Optimize small buffer copies

mwyrzykowski opened the pull request:
#### 032bebe8d4bd208a9a5a93a3797859589f6f9857
<pre>
[WebGPU] Optimize small buffer copies
<a href="https://bugs.webkit.org/show_bug.cgi?id=277647">https://bugs.webkit.org/show_bug.cgi?id=277647</a>
radar://133239762

Reviewed by Dan Glastonbury.

Avoid using shmem path for smaller sized buffers and textures as
cost of allocation is too high and given the allocation size was
page sized it may not having been saving memory for most copies.

* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBuffer.cpp:
(WebKit::RemoteBuffer::copyWithCopy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBuffer.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBuffer.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteQueue.cpp:
(WebKit::RemoteQueue::writeBufferWithCopy):
(WebKit::RemoteQueue::writeTextureWithCopy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteQueue.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteQueue.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteBufferProxy.cpp:
(WebKit::WebGPU::RemoteBufferProxy::copyFrom):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteBufferProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteQueueProxy.cpp:
(WebKit::WebGPU::RemoteQueueProxy::writeBuffer):
(WebKit::WebGPU::RemoteQueueProxy::writeTexture):

Canonical link: <a href="https://commits.webkit.org/291894@main">https://commits.webkit.org/291894@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f62a23f13a13ea001f33ad7e9356fb3ef5b3649f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/94291 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/13878 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3649 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/99303 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/44819 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/14178 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/22308 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71935 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29276 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/97293 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10531 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/85143 "Found 1 new API test failure: TestWebKitAPI.FullscreenVideoTextRecognition.TogglePlaybackInElementFullscreen (failure)") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52281 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/10220 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/44137 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80452 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2918 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101348 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21343 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/15557 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80941 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21595 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/81156 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80323 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20029 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24871 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/2247 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/14625 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21327 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/26506 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/21014 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24474 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22755 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUBuffer.h:
```diff
@@ -50,7 +50,8 @@ class Buffer : public RefCounted<Buffer>, public CanMakeWeakPtr<Buffer> {
 
     virtual void mapAsync(MapModeFlags, Size64 offset, std::optional<Size64>, CompletionHandler<void(bool)>&&) = 0;
-    virtual void getMappedRange(Size64 offset, std::optional<Size64>, Function<void(std::span<uint8_t>)>&&) = 0;
+    virtual void getMappedRange(Size64 offset, std::optional<Size64>, Function<void(std::pair<uint64_t, uint64_t>)>&&) = 0;
```
`size_t` for memory object sizes?

djg commented on Source/WebGPU/WebGPU/Device.mm:
```diff
@@ -209,6 +209,15 @@ static void captureFrame(id<MTLDevice> captureObject)
+bool Device::hasUnifiedMemory() const
+{
+#if CPU(X86_64)
+    return false;
```
Is `m_device.hasUnifiedMemory` returning `true` because of Intel Graphics?

mwyrzykowski replied:
yes

djg commented on Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteBufferProxy.cpp:
```diff
@@ -63,18 +65,60 @@ static bool offsetOrSizeExceedsBounds(size_t dataSize, WebCore::WebGPU::Size64 o
-    auto [data] = sendResult.takeReplyOr(std::nullopt);
+    auto [dataOffsetAndSize] = sendResult.takeReplyOr(std::make_pair(0ull, 0ull));
+    auto dataOffset = dataOffsetAndSize.first;
+    auto dataSize = dataOffsetAndSize.second;
```
Does destructuring assignment work here?
```
auto [dataOffset, dataSize] = sendResult.takeReplyOr(...);
```

mwyrzykowski replied:
oh most likely, I will give that a try

djg commented on Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteBufferProxy.cpp:
```diff
@@ -63,18 +65,60 @@ static bool offsetOrSizeExceedsBounds(size_t dataSize, WebCore::WebGPU::Size64 o
+    callback(std::make_pair(dataOffset, dataSize));
+}
+
+std::optional<std::span<uint8_t>> RemoteBufferProxy::mutableSpan()
```
What's the difference between a `std::optional<std::span<...>>` and a `std::span<...>` that is empty? Is it meaningful to distinguish between no buffer memory and the buffer being empty?

mwyrzykowski replied:
oh I think we could just use an empty span, good idea

djg commented on Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteQueueProxy.cpp:
```diff
@@ -69,12 +70,24 @@ void RemoteQueueProxy::onSubmittedWorkDone(CompletionHandler<void()>&& callback)
+            UNUSED_PARAM(sendResult);
+        }
+
+        auto newSize = static_cast<size_t>(size.value_or(source.size() - dataOffset));
```
Has dataOffset been validated to be < source.size() at this point?

mwyrzykowski replied:
`dataOffset` is always zero currently, because we pass `data.span().subspan(dataOffset, contentSize)` from `GPUQueue::writeBuffer`, but I should add some logic to ensure that or drop `dataOffset` from the parameter list

djg reviewed (approved):
:shipit:

mwyrzykowski committed:
[WebGPU] Optimize small buffer copies
https://bugs.webkit.org/show_bug.cgi?id=277647
radar://133239762

Reviewed by Dan Glastonbury.

Avoid using shmem path for smaller sized buffers and textures as
cost of allocation is too high and given the allocation size was
page sized it may not having been saving memory for most copies.

* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBuffer.cpp:
(WebKit::RemoteBuffer::copyWithCopy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBuffer.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBuffer.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteQueue.cpp:
(WebKit::RemoteQueue::writeBufferWithCopy):
(WebKit::RemoteQueue::writeTextureWithCopy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteQueue.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteQueue.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteBufferProxy.cpp:
(WebKit::WebGPU::RemoteBufferProxy::copyFrom):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteBufferProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteQueueProxy.cpp:
(WebKit::WebGPU::RemoteQueueProxy::writeBuffer):
(WebKit::WebGPU::RemoteQueueProxy::writeTexture):

Canonical link: https://commits.webkit.org/291894@main
